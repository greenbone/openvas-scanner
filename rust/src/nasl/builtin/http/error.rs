// SPDX-FileCopyrightText: 2025 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

use std::io;

use thiserror::Error;

#[derive(Debug, Error)]
pub enum HttpError {
    #[error("IO error during HTTP: {0}")]
    IO(io::ErrorKind),
    #[error("HTTP error: {0}")]
    H2(String),
    #[error("Handle ID {0} not found.")]
    HandleIdNotFound(i32),
}

impl From<io::Error> for HttpError {
    fn from(value: io::Error) -> Self {
        Self::IO(value.kind())
    }
}

impl From<h2::Error> for HttpError {
    fn from(value: h2::Error) -> Self {
        Self::H2(format!("{}", value))
    }
}
