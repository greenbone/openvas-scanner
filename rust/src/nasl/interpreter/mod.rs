mod assign;
mod call;
mod declare;
mod error;
mod forking_interpreter;
mod include;
mod loop_extension;
mod nasl_value;
mod register;

#[cfg(test)]
mod tests;

use std::collections::VecDeque;

use crate::nasl::{
    Code,
    error::Span,
    syntax::{
        Ident,
        grammar::{
            Array, ArrayAccess, Atom, Binary, BinaryOperatorKind, Block, Exit, Expr, If, Include,
            Return, Statement, Unary,
        },
    },
};

use InterpreterError as Error;
use InterpreterErrorKind as ErrorKind;
use error::ExprLocation;

use super::{
    ScanCtx, ScriptCtx,
    syntax::{LiteralKind, grammar::UnaryPrefixOperatorKind},
};

pub use error::{FunctionCallError, InterpreterError, InterpreterErrorKind};
pub use forking_interpreter::ForkingInterpreter;
pub use nasl_value::NaslValue;
pub use register::Register;

pub type Result<T = NaslValue, E = Error> = std::result::Result<T, E>;

#[derive(PartialEq, Eq)]
enum InterpreterState {
    Running,
    Finished,
}

impl InterpreterState {
    fn is_finished(&self) -> bool {
        matches!(self, Self::Finished)
    }
}

/// Represents the result of a function call (`value`) along with
/// the `Span` pointing to the identifier of the function that
/// resulted in this value originally.
#[derive(Clone)]
pub struct FunctionCallData {
    value: NaslValue,
    span: Span,
}

/// This type contains the necessary data to reproduce the execution
/// flow in the case of interpreter forks.
///
/// Its two variants are
///
/// 1. `Collecting`: This variant is used by any interpreter which is
///    currently executing normally.  In this mode, the result of any
///    function call performed by the interpreter within a single
///    top-level statement will be stored along with the `Span` at
///    which the function call took place (as a safeguard). The
///    variant also stores the original `Register` and statement index
///    in order to be able to "rewind" into the exact state before the
///    statement that caused the fork.
/// 2. `Restoring`: This variant is used by all interpreters which
///    were just created due to a fork.  In this mode, the interpreter
///    will not perform normal function calls, and will instead return
///    the first value in the `data` field in place of the function
///    call. This is done until the `data` field is exhausted. At that
///    point, execution proceeds normally.
#[derive(Clone)]
pub enum ForkReentryData {
    Collecting {
        data: Vec<FunctionCallData>,
        register: Register,
        stmt_index: usize,
    },
    Restoring {
        data: VecDeque<FunctionCallData>,
    },
}

impl ForkReentryData {
    fn drain(&mut self) -> VecDeque<FunctionCallData> {
        match self {
            Self::Collecting { data, .. } => data.drain(..).collect(),
            _ => unreachable!(),
        }
    }

    fn register(&self) -> &Register {
        match self {
            Self::Collecting { register, .. } => register,
            _ => unreachable!(),
        }
    }

    fn stmt_index(&self) -> usize {
        match self {
            Self::Collecting { stmt_index, .. } => *stmt_index,
            _ => unreachable!(),
        }
    }

    fn contains_fork(&self) -> bool {
        match self {
            Self::Collecting { data, .. } => data
                .iter()
                .any(|value| matches!(value.value, NaslValue::Fork(_))),
            _ => false,
        }
    }

    /// If in `Collecting` mode, remember the given result. Otherwise
    /// do nothing.
    pub(crate) fn try_collect(&mut self, value: NaslValue, span: &Span) {
        match self {
            ForkReentryData::Collecting { data, .. } => {
                data.push(FunctionCallData { value, span: *span })
            }
            ForkReentryData::Restoring { data: _ } => {}
        }
    }

    /// If in `Restoring` mode, remove and return the first stored
    /// result from the queue. Otherwise do nothing.
    pub(crate) fn try_restore(&mut self, span: &Span) -> Result<Option<NaslValue>, Error> {
        match self {
            Self::Restoring { data } => {
                if let Some(data) = data.pop_front() {
                    if *span != data.span {
                        return Err(ErrorKind::InvalidFork.with_span(&data.span));
                    }
                    Ok(Some(data.value))
                } else {
                    Ok(None)
                }
            }
            _ => Ok(None),
        }
    }

    fn create_forks(&mut self) -> Vec<Self> {
        let mut data = vec![self.drain()];
        loop {
            let (changed, new_data) = expand_first_fork(data);
            data = new_data;
            if !changed {
                break;
            }
        }
        data.into_iter()
            .map(|data| Self::Restoring { data })
            .collect()
    }

    fn new() -> Self {
        Self::Restoring {
            data: VecDeque::new(),
        }
    }

    fn collecting(register: Register, stmt_index: usize) -> Self {
        Self::Collecting {
            data: vec![],
            register,
            stmt_index,
        }
    }

    fn is_empty_or_collecting(&self) -> bool {
        match self {
            ForkReentryData::Collecting { .. } => true,
            ForkReentryData::Restoring { data } => data.is_empty(),
        }
    }
}

/// Expand the first occurrence of `NaslValue::Fork(...)` in the list of collected function
/// calls (i.e. any called function wants to fork), by returning one list per fork value.
///
/// For example (in pseudo-code):
/// expand_first_fork([[1, Fork(2, 3)]]) = [[1, 2], [1, 3]]
/// The boolean return value is true if any expansion took place or false if
/// no expansion took place (that is, if there was no `NaslValue::Fork` in the data).
fn expand_first_fork(
    data: Vec<VecDeque<FunctionCallData>>,
) -> (bool, Vec<VecDeque<FunctionCallData>>) {
    let first_fork = data[0]
        .iter()
        .enumerate()
        .filter_map(|(index, data)| {
            let FunctionCallData { value, span } = data;
            if let NaslValue::Fork(vals) = value {
                Some((index, vals.clone(), *span))
            } else {
                None
            }
        })
        .next();
    if first_fork.is_none() {
        return (false, data);
    }
    let first_fork = first_fork.unwrap();
    let data = data
        .into_iter()
        .flat_map(|d| expand_fork_at(d, first_fork.0, first_fork.1.clone(), first_fork.2))
        .collect();
    (true, data)
}

fn expand_fork_at(
    data: VecDeque<FunctionCallData>,
    index: usize,
    vals: Vec<NaslValue>,
    span: Span,
) -> Vec<VecDeque<FunctionCallData>> {
    vals.iter()
        .map(|val| {
            let mut data = data.clone();
            data[index] = FunctionCallData {
                value: val.clone(),
                span,
            };
            data
        })
        .collect()
}

pub struct Interpreter<'ctx> {
    pub(super) register: Register,
    pub(super) scan_ctx: &'ctx ScanCtx<'ctx>,
    pub(super) script_ctx: ScriptCtx,
    pub(super) fork_reentry_data: ForkReentryData,
    stmt_index: usize,
    state: InterpreterState,
}

impl<'ctx> Interpreter<'ctx> {
    /// Creates a new Interpreter
    fn new(register: Register, scan_ctx: &'ctx ScanCtx) -> Self {
        Interpreter {
            register,
            stmt_index: 0,
            scan_ctx,
            script_ctx: ScriptCtx::default(),
            fork_reentry_data: ForkReentryData::new(),
            state: InterpreterState::Running,
        }
    }

    pub(super) async fn execute_statement(
        &mut self,
        stmt: Option<&Statement>,
    ) -> Option<Result<NaslValue, Error>> {
        self.initialize_fork_data();
        self.stmt_index += 1;
        match stmt {
            Some(stmt) => {
                let result = self.resolve(stmt).await;
                if matches!(result, Ok(NaslValue::Exit(_))) {
                    self.state = InterpreterState::Finished;
                }
                Some(result)
            }
            None => {
                self.state = InterpreterState::Finished;
                None
            }
        }
    }

    pub(super) async fn resolve(&mut self, statement: &Statement) -> Result<NaslValue, Error> {
        use Statement::*;
        match statement {
            NoOp => Ok(NaslValue::Null),
            Continue => Ok(NaslValue::Continue),
            Break => Ok(NaslValue::Break),
            ExprStmt(expr) => self.resolve_expr(expr).await,
            Block(block) => self.resolve_block(block).await,
            VarScopeDecl(var_scope_decl) => self.resolve_var_scope_decl(var_scope_decl),
            FnDecl(fn_decl) => self.resolve_fn_decl(fn_decl),
            If(if_) => self.resolve_if(if_).await,
            While(while_) => Box::pin(self.resolve_while(while_)).await,
            Repeat(repeat) => Box::pin(self.resolve_repeat(repeat)).await,
            ForEach(foreach) => Box::pin(self.resolve_foreach(foreach)).await,
            For(for_) => Box::pin(self.resolve_for(for_)).await,
            Exit(exit) => self.resolve_exit(exit).await,
            Include(include_) => self.resolve_include(include_).await,
            Return(return_) => self.resolve_return(return_).await,
        }
    }

    pub(super) async fn resolve_expr(&mut self, expr: &Expr) -> Result {
        match expr {
            Expr::Atom(atom) => Box::pin(self.resolve_atom(atom)).await,
            Expr::Binary(binary) => Box::pin(self.resolve_binary(binary)).await,
            Expr::Unary(unary) => Box::pin(self.resolve_unary(unary)).await,
            Expr::Assignment(assignment) => Box::pin(self.resolve_assignment(assignment)).await,
        }
    }

    pub async fn collect_exprs(
        &mut self,
        exprs: impl Iterator<Item = &Expr>,
    ) -> Result<Vec<NaslValue>> {
        let mut vals = vec![];
        for array_access in exprs {
            vals.push(self.resolve_expr(array_access).await?);
        }
        Ok(vals)
    }

    async fn resolve_atom(&mut self, atom: &Atom) -> Result {
        match atom {
            Atom::Literal(literal) => Ok(match &literal.kind {
                LiteralKind::String(s) => NaslValue::String(s.clone()),
                LiteralKind::Data(data) => NaslValue::Data(data.clone()),
                LiteralKind::Number(num) => NaslValue::Number(*num),
                LiteralKind::IPv4Address(ipv4_addr) => NaslValue::String(ipv4_addr.to_string()),
                LiteralKind::Null => NaslValue::Null,
                LiteralKind::Boolean(b) => NaslValue::Boolean(*b),
                LiteralKind::AttackCategory(a) => NaslValue::AttackCategory(*a),
                LiteralKind::FCTAnonArgs => NaslValue::Array(self.register.positional().to_vec()),
            }),
            Atom::FnCall(call) => self.resolve_fn_call(call).await,
            Atom::Ident(ident) => Ok(self.resolve_var(ident)?.clone()),
            Atom::Array(array) => self.resolve_array(array).await,
            Atom::ArrayAccess(array_access) => self.resolve_array_access(array_access).await,
            Atom::Increment(increment) => self.resolve_increment(increment).await,
        }
    }

    fn resolve_var(&self, ident: &Ident) -> Result {
        let var = self.register.get(ident.to_str());
        if let Some(var) = var {
            Ok(self
                .register
                .get_val(var)
                .as_value()
                .map_err(|e| e.with_span(&ident))?
                .clone())
        } else {
            Ok(NaslValue::Null)
        }
    }

    async fn resolve_array_access(&mut self, array_access: &ArrayAccess) -> Result {
        let lhs = self.resolve_expr(&array_access.lhs_expr).await?;
        let index = self.resolve_expr(&array_access.index_expr).await?;
        lhs.index(index)
            .map_err(|e| e.into_error(&array_access.lhs_expr, &array_access.index_expr))
    }

    async fn resolve_unary(&mut self, unary: &Unary) -> Result {
        let rhs = self.resolve_expr(&unary.rhs).await?;
        match unary.op.kind {
            UnaryPrefixOperatorKind::Plus => Ok(rhs),
            UnaryPrefixOperatorKind::Minus => rhs.neg(),
            UnaryPrefixOperatorKind::Bang => rhs.not(),
            UnaryPrefixOperatorKind::Tilde => rhs.bitwise_not(),
        }
        .map_err(|e| e.with_span(&*unary.rhs))
    }

    async fn resolve_binary(&mut self, binary: &Binary) -> Result {
        use BinaryOperatorKind::*;
        let lhs = self.resolve_expr(&binary.lhs).await?;
        // Short circuit
        if binary.op.kind == AmpersandAmpersand && !lhs.convert_to_boolean() {
            return Ok(NaslValue::Boolean(false));
        }
        if binary.op.kind == PipePipe && lhs.convert_to_boolean() {
            return Ok(NaslValue::Boolean(true));
        }
        let rhs = self.resolve_expr(&binary.rhs).await?;
        let eval = || match binary.op.kind {
            Plus => Ok(lhs.add(rhs)),
            Minus => Ok(lhs.sub(rhs)),
            Star => lhs.mul(rhs),
            Slash => lhs.div(rhs),
            Percent => lhs.rem(rhs),
            StarStar => lhs.pow(rhs),
            LessLess => lhs.shl(rhs),
            GreaterGreater => lhs.shr(rhs),
            GreaterGreaterGreater => lhs.shr_unsigned(rhs),
            Less => lhs.lt(rhs),
            LessEqual => lhs.le(rhs),
            Greater => lhs.gt(rhs),
            GreaterEqual => lhs.ge(rhs),
            Ampersand => lhs.bitand(rhs),
            Pipe => lhs.bitor(rhs),
            Caret => lhs.bitxor(rhs),
            AmpersandAmpersand => Ok(lhs.and(rhs)),
            PipePipe => Ok(lhs.or(rhs)),
            EqualTilde => Ok(NaslValue::Boolean(lhs.match_regex(rhs)?)),
            BangTilde => Ok(NaslValue::Boolean(!lhs.match_regex(rhs)?)),
            GreaterLess => Ok(NaslValue::Boolean(lhs.match_string(rhs)?)),
            GreaterBangLess => Ok(NaslValue::Boolean(!lhs.match_string(rhs)?)),
            EqualEqual => Ok(NaslValue::Boolean(lhs == rhs)),
            BangEqual => Ok(NaslValue::Boolean(lhs != rhs)),
        };
        eval().map_err(|err| err.into_error(&binary.lhs, &binary.rhs))
    }

    async fn resolve_array(&mut self, array: &Array) -> Result {
        Ok(NaslValue::Array(
            self.collect_exprs(array.items.items.iter()).await?,
        ))
    }

    async fn resolve_exit(&mut self, exit: &Exit) -> Result {
        let rc = Box::pin(self.resolve_expr(&exit.expr)).await?;
        match rc {
            NaslValue::Number(rc) => Ok(NaslValue::Exit(rc)),
            _ => Err(ErrorKind::NonNumericExitCode(rc).with_span(&exit.expr)),
        }
    }

    async fn resolve_return(&mut self, return_: &Return) -> Result {
        let rc = if let Some(ref expr) = return_.expr {
            self.resolve_expr(expr).await?
        } else {
            NaslValue::Null
        };
        Ok(NaslValue::Return(Box::new(rc)))
    }

    async fn resolve_if(
        &mut self,
        If {
            if_branches,
            else_branch,
        }: &If,
    ) -> Result<NaslValue, Error> {
        for (condition, block) in if_branches.iter() {
            let val = self.resolve_expr(condition).await?;
            if val.convert_to_boolean() {
                return self.resolve_block(block).await;
            }
        }
        if let Some(block) = else_branch {
            return self.resolve_block(block).await;
        }
        Ok(NaslValue::Null)
    }

    pub(crate) async fn resolve_block(&mut self, block: &Block<Statement>) -> Result {
        self.register.create_child();
        for stmt in block.items.iter() {
            match Box::pin(self.resolve(stmt)).await {
                Ok(x) => {
                    if matches!(
                        x,
                        NaslValue::Exit(_)
                            | NaslValue::Return(_)
                            | NaslValue::Break
                            | NaslValue::Continue
                    ) {
                        self.register.drop_last();
                        return Ok(x);
                    }
                }
                Err(e) => return Err(e),
            }
        }
        self.register.drop_last();
        // currently blocks return null
        Ok(NaslValue::Null)
    }

    async fn resolve_include(&mut self, include: &Include) -> Result {
        let loader = self.scan_ctx.loader();
        let code = Code::load(loader, &include.path)
            .map_err(|e| ErrorKind::LoadError(e).with_span(&include.span))?
            .parse();
        let file = code.file().clone();
        let ast = code
            .result()
            .map_err(|errs| Error::include_syntax_error(errs, file))?;
        let mut inter = ForkingInterpreter::new(ast, self.register.clone(), self.scan_ctx);
        Box::pin(inter.execute_all()).await?;
        self.register = inter.register().clone();
        Ok(NaslValue::Null)
    }

    pub(crate) fn make_forks(mut self) -> Vec<Interpreter<'ctx>> {
        let forks = self.fork_reentry_data.create_forks();
        let register = self.fork_reentry_data.register();
        let stmt_index = self.fork_reentry_data.stmt_index();
        forks
            .into_iter()
            .map(|fork| self.make_fork(fork, register, stmt_index))
            .collect()
    }

    fn make_fork(
        &self,
        fork_reentry_data: ForkReentryData,
        register: &Register,
        stmt_index: usize,
    ) -> Interpreter<'ctx> {
        Self {
            register: register.clone(),
            stmt_index,
            scan_ctx: self.scan_ctx,
            script_ctx: ScriptCtx::default(),
            fork_reentry_data,
            state: InterpreterState::Running,
        }
    }

    pub(crate) fn wants_to_fork(&self) -> bool {
        self.fork_reentry_data.contains_fork()
    }

    pub(crate) fn initialize_fork_data(&mut self) {
        if self.fork_reentry_data.is_empty_or_collecting() {
            self.fork_reentry_data =
                ForkReentryData::collecting(self.register.clone(), self.stmt_index);
        }
    }

    pub(crate) fn is_finished(&self) -> bool {
        self.state.is_finished()
    }

    pub fn register(&self) -> &Register {
        &self.register
    }
}
