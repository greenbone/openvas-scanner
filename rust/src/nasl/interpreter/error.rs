// SPDX-FileCopyrightText: 2023 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

use crate::nasl::NaslValue;
use crate::nasl::code::SourceFile;
use crate::nasl::error::{AsCodespanError, Level, Span, Spanned, emit_errors};
use crate::nasl::syntax::grammar::Expr;
use crate::nasl::syntax::{CharIndex, ParseError};
use crate::nasl::syntax::{Ident, LoadError};
use crate::nasl::utils::error::FnError;
use codespan_reporting::files::SimpleFile;
use thiserror::Error;

#[derive(Debug, Error)]
/// An error that occurred while calling a function
#[error("Error while calling function '{function}': {kind}")]
pub struct FunctionCallError {
    /// Name of the function
    pub function: String,
    /// Kind of error
    #[source]
    pub kind: FnError,
}

impl FunctionCallError {
    /// Creates a new FunctionError
    pub fn new(function: &str, kind: FnError) -> Self {
        Self {
            function: function.to_owned(),
            kind,
        }
    }

    fn retryable(&self) -> bool {
        self.kind.retryable()
    }
}

#[derive(Debug, Error)]
/// Is used to represent an error while interpreting
#[error("{} {kind}", self.format_origin())]
pub struct InterpreterError {
    /// The kind of error that occurred.
    #[source]
    pub kind: InterpreterErrorKind,
    pub span: Span,
}

impl InterpreterError {
    fn format_origin(&self) -> String {
        // TODO
        String::new()
    }

    pub fn retryable(&self) -> bool {
        match &self.kind {
            InterpreterErrorKind::LoadError(LoadError::Retry(_)) => true,
            InterpreterErrorKind::FunctionCallError(e) => e.retryable(),
            _ => false,
        }
    }
}

#[derive(Debug, Error)]
/// Is used to give hints to the user how to react on an error while interpreting
pub enum InterpreterErrorKind {
    /// When returned context is a function when a value is required.
    #[error("Expected a value but got a function.")]
    FunctionExpectedValue,
    /// When returned context is a value when a function is required.
    #[error("Expected a function but got a value.")]
    ValueExpectedFunction,
    /// Regex parsing went wrong.
    #[error("Invalid regular expression: {0}")]
    InvalidRegex(String),
    /// A SyntaxError while including another script
    #[error("{0}")]
    IncludeSyntaxError(IncludeSyntaxError),
    /// Syntax errors in the script.
    #[error("Syntax errors occurred.")]
    SyntaxError(Vec<ParseError>),
    /// When the given function is undefined.
    #[error("Undefined function: {0}")]
    UndefinedFunction(String),
    /// A LoadError occurred
    #[error("{0}")]
    LoadError(LoadError),
    /// An error occurred while calling a built-in function.
    #[error("{0}")]
    FunctionCallError(FunctionCallError),
    /// A script tried to fork in a way that caused
    /// forks to run into different branches of
    /// the original statement which caused the fork.
    #[error(
        "Invalid fork. The interpreter forked in a position which was not reached by the created forks."
    )]
    InvalidFork,
    #[error("Expected a string.")]
    ExpectedString,
    #[error("Expected a number.")]
    ExpectedNumber,
    #[error("Assignment to undefined variable: {0}")]
    AssignmentToUndefinedVar(Ident),
    #[error("Array out of range for index: {0}")]
    ArrayOutOfRange(usize),
    #[error("Negative index into array: {0}")]
    NegativeIndex(i64),
    #[error("Dict key does not exist: {0}")]
    DictKeyDoesNotExist(String),
    #[error("Expected array or dict.")]
    ArrayOrDictExpected,
    #[error("Tried to exit with non-numeric exit code {0}.")]
    NonNumericExitCode(NaslValue),
}

impl InterpreterErrorKind {
    pub(crate) fn with_span(self, s: &impl Spanned) -> InterpreterError {
        InterpreterError {
            kind: self,
            span: s.span(),
        }
    }
}

impl InterpreterError {
    pub(crate) fn syntax_error(errs: Vec<ParseError>) -> Self {
        Self {
            // fake value since we don't need the span in this case
            span: Span::new(CharIndex(usize::MAX - 1), CharIndex(usize::MAX)),
            kind: InterpreterErrorKind::SyntaxError(errs),
        }
    }

    pub(crate) fn include_syntax_error(
        errs: Vec<ParseError>,
        file: SimpleFile<String, String>,
    ) -> Self {
        Self {
            // fake value since we don't need the span in this case
            span: Span::new(CharIndex(usize::MAX - 1), CharIndex(usize::MAX)),
            kind: InterpreterErrorKind::IncludeSyntaxError(IncludeSyntaxError { errs, file }),
        }
    }
}

#[derive(Debug)]
// TODO
pub struct IncludeSyntaxError {
    pub file: SourceFile,
    pub errs: Vec<ParseError>,
}

// TODO Get rid of this once we have a proper implementation of spans
// for InterpreterError as well.
impl std::fmt::Display for IncludeSyntaxError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        emit_errors(&self.file, self.errs.iter().cloned(), Level::Error);
        write!(f, "Syntax error while including file.")
    }
}

impl InterpreterError {
    pub(crate) fn new(kind: InterpreterErrorKind, span: Span) -> Self {
        Self { kind, span }
    }
}

impl AsCodespanError for &InterpreterError {
    fn span(&self) -> Span {
        self.span
    }

    fn message(&self) -> String {
        self.kind.to_string()
    }
}

pub(super) struct ExprError {
    pub kind: InterpreterErrorKind,
    pub location: ExprLocation,
}

pub(super) enum ExprLocation {
    Lhs,
    Rhs,
    Both,
}

impl ExprError {
    pub fn lhs(kind: InterpreterErrorKind) -> Self {
        Self {
            kind,
            location: ExprLocation::Lhs,
        }
    }

    pub fn rhs(kind: InterpreterErrorKind) -> Self {
        Self {
            kind,
            location: ExprLocation::Rhs,
        }
    }

    pub fn both(kind: InterpreterErrorKind) -> Self {
        Self {
            kind,
            location: ExprLocation::Both,
        }
    }

    pub(crate) fn into_error(self, lhs_expr: &Expr, rhs_expr: &Expr) -> InterpreterError {
        let span = match self.location {
            ExprLocation::Lhs => lhs_expr.span(),
            ExprLocation::Rhs => rhs_expr.span(),
            ExprLocation::Both => lhs_expr.span().join(rhs_expr.span()),
        };
        InterpreterError::new(self.kind, span)
    }
}
