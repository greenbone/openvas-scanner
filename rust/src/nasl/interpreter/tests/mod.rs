// SPDX-FileCopyrightText: 2025 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

use codespan_reporting::files::SimpleFile;

use crate::nasl::error::emit_errors_str;
use crate::nasl::{NaslResult, test_utils::TestBuilder};

mod control_flow;
mod description;
mod local_var;
mod retry;

pub fn interpret(code: &str) -> Vec<NaslResult> {
    let mut t = TestBuilder::default();
    t.run_all(code);
    t.results()
}

pub fn interpret_err(file_name: &str, code: &str) -> String {
    let mut t = TestBuilder::default();
    t.run_all(code);
    let err = t
        .interpreter_results()
        .into_iter()
        .find(|result| result.is_err())
        .unwrap()
        .unwrap_err();
    emit_errors_str(
        &SimpleFile::new(file_name.to_string(), code.to_string()),
        std::iter::once(&err),
    )
}

#[macro_export]
macro_rules! interpreter_test_ok {
    ($name: ident, $code: literal, $($expected: expr),* $(,)?) => {
        #[test]
        fn $name() {
            let mut results = $crate::nasl::interpreter::tests::interpret($code);
            let mut count = 0;
            $(
                count += 1;
                let result = results.remove(0).unwrap();
                let expected = $expected.to_nasl_result().unwrap();
                assert_eq!(expected, result, "mismatch in result #{count}. Expected: {expected:?}, found {result:?}.");
            )*
            assert_eq!(results.len(), 0);
        }
    };
}

#[macro_export]
macro_rules! interpreter_test_err {
    ($name: ident, $code: literal) => {
        #[test]
        fn $name() {
            insta::assert_snapshot!($crate::nasl::interpreter::tests::interpret_err(
                stringify!($name),
                $code
            ));
        }
    };
}

mod operator {
    use crate::interpreter_test_ok;
    use crate::nasl::test_prelude::*;

    interpreter_test_ok!(numeric_plus, "1 + 2;", 3);
    interpreter_test_ok!(cast_to_string_middle_plus, "1+\"\"+2;", "12");
    interpreter_test_ok!(cast_to_string_end_plus, "1+2+\"\";", "3");
    interpreter_test_ok!(cast_to_string_end_plus_4, "1+2+\"\" + 4;", "34");
    interpreter_test_ok!(cast_to_string_minus, "11-\"1\";", "1");
    interpreter_test_ok!(string_plus, "\"hello \" + \"world!\";", "hello world!");
    interpreter_test_ok!(string_minus, "\"hello \" - 'o ';", "hell");
    interpreter_test_ok!(data_plus, "'hello ' + 'world!';", "hello world!".as_bytes());
    interpreter_test_ok!(data_minus, "'hello ' - 'o ';", "hell".as_bytes());

    interpreter_test_ok!(cast_to_data_middle_plus, "1+''+2;", "12".as_bytes());
    interpreter_test_ok!(cast_to_data_end_plus, "1+2+'';", "3".as_bytes());
    interpreter_test_ok!(cast_to_data_end_plus_4, "1+2+'' + 4;", "34".as_bytes());
    interpreter_test_ok!(cast_to_data_minus, "11-'1';", "1".as_bytes());
    interpreter_test_ok!(numeric_minus, "1 - 2;", -1);
    interpreter_test_ok!(multiplication, "1*2;", 2);
    interpreter_test_ok!(division, "512/2;", 256);
    interpreter_test_ok!(modulo, "512%2;", 0);
    interpreter_test_ok!(left_shift, "512 << 2;", 2048);
    interpreter_test_ok!(right_shift, "512 >> 2;", 128);
    interpreter_test_ok!(unsigned_right_shift, "-2 >>> 2;", 1073741823);
    interpreter_test_ok!(and, "-2 & 2;", 2);
    interpreter_test_ok!(or, "-2 | 2;", -2);
    interpreter_test_ok!(xor, "-2 ^ 2;", -4);
    interpreter_test_ok!(pow, "2 ** 2;", 4);
    interpreter_test_ok!(not, "~2;", -3);
    interpreter_test_ok!(r_match, "'hello' =~ 'hell';", true);
    interpreter_test_ok!(r_not_match, "'hello' !~ 'hell';", false);
    interpreter_test_ok!(contains, "'hello' >< 'hell';", true);
    interpreter_test_ok!(not_contains, "'hello' >!< 'hell';", false);
    interpreter_test_ok!(bool_not, "!23;", false);
    interpreter_test_ok!(bool_not_reverse, "!0;", true);
    interpreter_test_ok!(bool_and, "1 && 1;", true);
    interpreter_test_ok!(bool_or, "1 || 0;", true);
    interpreter_test_ok!(equals_data, "'1' == '1';", true);
    interpreter_test_ok!(equals_data_unequal, "'1' == '2';", false);
    interpreter_test_ok!(equals_string, "\"1\" == \"1\";", true);
    interpreter_test_ok!(equals_string_unequal, "\"1\" == \"2\";", false);
    interpreter_test_ok!(equals_number, "1 != 1;", false);
    interpreter_test_ok!(not_equals_data, "'1' != '1';", false);
    interpreter_test_ok!(not_equals_data_unequal, "'1' != '2';", true);
    interpreter_test_ok!(not_equals_string, "\"1\" != \"1\";", false);
    interpreter_test_ok!(not_equals_string_unequal, "\"1\" != \"2\";", true);
    interpreter_test_ok!(not_equals_number, "1 != 1;", false);
    interpreter_test_ok!(greater, "1 > 0;", true);
    interpreter_test_ok!(greater2, "1 < 0;", false);
    interpreter_test_ok!(less, "1 < 2;", true);
    interpreter_test_ok!(less2, "1 < 0;", false);
    interpreter_test_ok!(greater_equal, "1 >= 1;", true);
    interpreter_test_ok!(greater_equal_2, "1 >= 2;", false);
    interpreter_test_ok!(less_equal, "1 <= 1;", true);
    interpreter_test_ok!(less_equal2, "1 <= 0;", false);

    interpreter_test_ok!(
        x_gonna_give_it_to_ya,
        "
            a = 0;
            function test() {
                a++;
                return \"foo\";
            }
            test('hi') x 5;
            a;
        ",
        0,
        NaslValue::Null,
        "foo",
        5,
    );
}

use crate::nasl::test_prelude::*;

interpreter_test_ok!(block, "{ 3; 4; }", NaslValue::Null);
interpreter_test_ok!(
    block_scope,
    "a = 1; b = 2; { local_var a; a = 3; b = 4; } a; b;",
    1,
    2,
    NaslValue::Null,
    1,
    4,
);

interpreter_test_ok!(
    array_boolean,
    "a = 1; if ([]) { a = 2; }; a;",
    1,
    NaslValue::Null,
    NaslValue::Null,
    2
);

interpreter_test_ok!(
    null_boolean,
    "a = 1; if (b) { a = 2; }; a;",
    1,
    NaslValue::Null,
    NaslValue::Null,
    1
);

interpreter_test_ok!(
    integer_boolean,
    "a = 1; if (a) { a = 2; }; a;",
    1,
    NaslValue::Null,
    NaslValue::Null,
    2
);

interpreter_test_ok!(
    integer_boolean_zero,
    "a = 0; if (a) { a = 2; }; a;",
    0,
    NaslValue::Null,
    NaslValue::Null,
    0
);

interpreter_test_ok!(
    string_boolean,
    "a = \"foo\"; if (a) { a = 2; }; a;",
    "foo",
    NaslValue::Null,
    NaslValue::Null,
    2
);

interpreter_test_ok!(
    string_boolean_empty,
    "a = \"\"; if (a) { a = 2; }; a;",
    "",
    NaslValue::Null,
    NaslValue::Null,
    "",
);

interpreter_test_ok!(
    string_boolean_zero,
    "a = \"0\"; if (a) { a = 2; }; a;",
    "0",
    NaslValue::Null,
    NaslValue::Null,
    "0",
);

interpreter_test_ok!(
    non_int_array_index,
    "a = [1, 2, 3]; a['hello'];",
    NaslValue::Array(vec![1.into(), 2.into(), 3.into()]),
    NaslValue::Null,
);

interpreter_test_err!(nonexistent_variable, "a += 12;");

interpreter_test_err!(function_instead_of_variable, "function foo() { } a = foo;");
interpreter_test_err!(variable_instead_of_function, "foo = 3; foo();");

interpreter_test_err!(invalid_regex, r#"a = "hello world"; a =~ "[";"#);

interpreter_test_err!(undefined_fn, "foo();");

interpreter_test_err!(expected_str, "a =~ 5;");
interpreter_test_err!(expected_array, "a = 5; a[3];");

interpreter_test_err!(array_out_of_range, "a = [1, 2, 3]; a[3];");
interpreter_test_err!(negative_array_index, "a = [1, 2, 3]; a[-3];");
