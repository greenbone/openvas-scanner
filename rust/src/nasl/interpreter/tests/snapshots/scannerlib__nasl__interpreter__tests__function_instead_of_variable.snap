---
source: src/nasl/interpreter/tests/mod.rs
expression: "crate :: nasl :: interpreter :: tests ::\ninterpret_err(stringify! (function_instead_of_variable),\n\"function foo() { } a = foo;\")"
---
error: Expected a value but got a function.
  ┌─ function_instead_of_variable:1:24
  │
1 │ function foo() { } a = foo;
  │                        ^^^ Expected a value but got a function.
